keeper/src/main/scala/zio/keeper/consensus/JGroups.scala

```diff
@@ -0,0 +1,147 @@
+package zio.keeper.consensus
+
+import zio.{ Chunk, Has, IO, Tagged, UIO, ZIO, ZLayer, clock, logging }
+import zio.clock.Clock
+import zio.duration.Duration
+import zio.keeper.Error
+import zio.keeper.membership.{ ByteCodec, TaggedCodec }
+import zio.stream.ZSink
+import upickle.default._
+import zio.logging.Logging
+
+/**
+ *  Implemented by user
+ */
+object ReceiverAdapter {
+
+  trait Service {
+    def receive(data: Chunk[Byte]): IO[Error, Unit]
+
+    def setState(data: Chunk[Byte]): IO[Error, Unit]
+
+    def getState: IO[Error, Chunk[Byte]]
+  }
+
+}
+
+object JGroups {
+
+  sealed private trait JGroupsProtocol
+  private case object StateRequest                       extends JGroupsProtocol
+  private case class StateResponse(payload: Chunk[Byte]) extends JGroupsProtocol
+  private case class UserMessage(payload: Chunk[Byte])   extends JGroupsProtocol
+
+  private val stateRequestCodec = ByteCodec.fromReadWriter(
+    implicitly[ReadWriter[Array[Byte]]].bimap[StateRequest.type](_ => Array(1, 1), _ => StateRequest)
+  )
+
+  private val StateResponseCodec = ByteCodec.fromReadWriter(
+    implicitly[ReadWriter[Array[Byte]]]
+      .bimap[StateResponse](_.payload.toArray, bA => StateResponse(Chunk.fromArray(bA)))
+  )
+
+  private val UserMessageCodec = ByteCodec.fromReadWriter(
+    implicitly[ReadWriter[Array[Byte]]].bimap[UserMessage](_.payload.toArray, bA => UserMessage(Chunk.fromArray(bA)))
+  )
+
+  implicit private val protocolCodec: TaggedCodec[JGroupsProtocol] =
+    TaggedCodec.instance[JGroupsProtocol](
+      {
+        case _: StateRequest.type => 21
+        case _: StateResponse     => 22
+        case _: UserMessage       => 23
+      }, {
+        case 21 => stateRequestCodec.asInstanceOf[ByteCodec[JGroupsProtocol]]
+        case 22 => StateResponseCodec.asInstanceOf[ByteCodec[JGroupsProtocol]]
+        case 23 => UserMessageCodec.asInstanceOf[ByteCodec[JGroupsProtocol]]
+      }
+    )
+
+  trait Service {
+    def send(data: Chunk[Byte]): IO[Error, Unit]
+  }
+
+  val live: ZLayer[Consensus, Error, JGroups] =
+    ZLayer.fromService[Consensus.Service, JGroups.Service] { c =>
+      new Service {
+        override def send(data: Chunk[Byte]): IO[Error, Unit] =
+          for {
+            payload <- TaggedCodec.write[JGroupsProtocol](UserMessage(data))
+            _       <- c.broadcast(payload)
+          } yield ()
+      }
+    }
+
+  private val msgLoop: ZIO[Consensus with ReceiverAdapter with Logging, Error, Unit] =
+    for {
+      c      <- ZIO.access[Consensus](_.get[Consensus.Service])
+      ra     <- ZIO.access[ReceiverAdapter](_.get[ReceiverAdapter.Service])
+      selfId <- c.selfNode
+      _ <- c.receive.tap { rawMsg =>
+            for {
```

Comment:
Nitpick: Dropping this for-comprehension in favor of plain `flatMap` would increase readability here.